### PR TITLE
Add funnel visualization

### PR DESCRIPTION
**Pull request by tonyjiangh**
Issue
---
https://github.com/getredash/redash/issues/2139

Implementation
---
- Table + Containers
- D3 for DOM manipulation 
- Bootstrap for formatting mostly
- Code structure based on sunburst

Possible alternatives: [svg with d3](https://codepen.io/tonyjiangh/pen/xYWbya)

Task lists
---
- [x] Add basic funnel visualization
- [x] Fix color styles
- [x] Optimize edit page
- [x] Test some corner cases
   - [x] Long list of steps
   - [x] Long step names
   - [x] Big values
   - [x] Small ratio for the last step
   - [x] Extreme resizing
- [x] Maybe some hover handling if necessary (for long step names?)
- ~~Documentation~~
  - Different repository

Screenshot
---
**Auto sorted setting**
image

**Order specified setting**
image

Updates
----
- Initial commit
  - Add basic funnel visualization
  - QueryResult has to have `steps` and `values` column
  - Auto sort records using value column in decreasing order
  - Auto calculate % of max and % of previous
- Added @ 2018/2/23(JP)
  - Allow choosing column to use for step names and for value
  - Allow setting display name for the funnel chart(table) header for the step columna and the value column
  - Allow choosing column to use for ordering
  - Use chart color `Cyan`
  - Normalize date column for steps display
  - Stop rendering when 1. value column is NaN or not number, 2. selected column name is invalid
- Added @ 2018/2/26~27
  - Silece top 100 rows for display
  - Handled long step name with ellipsis text-overflow and hovered full step name.
  - Some styling adjustment

### CODE REVIEW

**Comment by kocsmy** [comment]
Regarding visualization colors: Can we use the built-in visualization colors instead of introducing new colors? (see merged PR: https://github.com/getredash/redash/pull/2282)

**Comment by kravets-levko** [comment]
Hi @tonyjiangh! Looks really good. How about adding some options for this viz? It would be nice to choose table columns to display (I mean middle column with bars, "Total" and "Previous" columns). Also it would be great (but not necessary at this stage) to map result columns (to replace hard-coded `steps` and `values` ones).

**Comment by tonyjiangh** [comment]
@kocsmy Thanks for reminding me. That's one thing I forgot. I'll fix that.

@kravets-levko Yes, I do have plans for that. I'll add some explanation and screenshot after adding a commit.

File: client/app/visualizations/funnel/index.js [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+    drawFunnel(data); // draw funnel
+  }
+
+  function removeVisualization() {
+    vis.selectAll('table').remove();
+  }
+
+  function prepareData(queryData, stepCol, valCol) {
+    // Column validity
+    const sortedData = sortBy(queryData, valCol).reverse();
```
[the remark this answers is not in the document]
Comment (tonyjiangh) [reply client/app/visualizations/funnel/index.js]: Thanks for the advices.
Yeah, an option sounds better. I'll add that.

**Comment by tonyjiangh** [comment]
@arikfr 
Hi, I think I'm ready for review.
Something I'm not quite sure here and want some opinions:
- the codeclimate is reporting long method, high cognitive complexity, and similar code block which happends for many existed code as well. Should I fix them?
- No pagination here for the funnel table now, so I'm limiting the max-rows to be 100. We could also set no-limitation or add an option for it. (Are there really use-case for funnel data of more than 100 steps?)

File: client/app/visualizations/funnel/index.js [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+  ngModule.directive('funnelEditor', funnelEditor);
+
+  ngModule.config((VisualizationProvider) => {
+    const renderTemplate =
+      '<funnel-renderer options="visualization.options" query-result="queryResult"></funnel-renderer>';
+
+    const editTemplate = '<funnel-editor></funnel-editor>';
+    const defaultOptions = {
+      stepCol: { colName: '', dispAs: 'Steps' },
+      valueCol: { colName: '', dispAs: 'Value' },
```
Comment (arikfr) [inline client/app/visualizations/funnel/index.js]: One small thing, let's rename `dispAs` to `displayAs`.

**Comment by arikfr** [comment]
I missed your question about CodeClimate -- no, it's safe to ignore those. 

Also I think capping steps at 100 is reasonable, as probably no one will have more than 10 or so.

**Comment by arikfr** [comment]
💃

**Comment by eastpeak** [comment]
Just a quick spelling note: "Dispaly" should be "Display" in "Step Column Dispaly Name" and "Funnel Value Column Dispaly Name."

**Comment by rob-01** [comment]
Just a quick spelling note: "Dispaly" should be "Display" in "Step Column Dispaly Name" and "Funnel Value Column Dispaly Name."

**Comment by harshgupta29** [comment]
I have steps and values as a query result but when I try to make the funnel, it doesn't appear on the visualization panel. What to do???
